nearcache.go

```diff
@@ -0,0 +1,690 @@
+	}
+	return rid, nil
+}
+
+func (rs *nearCacheRecordStore) nextReservationID() int64 {
+	return atomic.AddInt64(&rs.reservationID, 1)
+}
+
+func (rs *nearCacheRecordStore) reserveForWriteUpdate(key interface{}, keyData serialization.Data, reservationID int64) (*nearCacheRecord, error) {
+	rs.recordsMu.Lock()
+	defer rs.recordsMu.Unlock()
+	rec, ok := rs.records[key]
+	if !ok {
+		rec, err := rs.createRecord(nil)
+		if err != nil {
+			return nil, err
+		}
+		rec.SetReservationID(reservationID)
+		// initInvalidationMetaData(record, key, keyData);
+	}
+	if rec.reservationID == nearCacheRecordReadPermitted {
+		rec.SetReservationID(reservationID)
+		return rec, nil
+	}
+	// 3. If this record is a previously reserved one, delete it.
+	// Reasoning: CACHE_ON_UPDATE mode has different characteristics
+	// than INVALIDATE mode when updating local near-cache. During
+	// update, if CACHE_ON_UPDATE finds a previously reserved
+	// record, that record is deleted. This is different from
+	// INVALIDATE mode which doesn't delete previously reserved
+	// record and keeps it as is. The reason for this deletion
+	// is: concurrent reservation attempts. If CACHE_ON_UPDATE
+	// doesn't delete previously reserved record, indefinite
+	// read of stale value situation can be seen. Since we
+	// don't apply invalidations which are sent from server
+	// to near-cache if the source UUID of the invalidation
+	// is same with the end's UUID which has near-cache on
+	// it (client or server UUID which has near cache on it).
+	return nil, nil
+}
+
+func (rs *nearCacheRecordStore) reserveForReadUpdate(key interface{}, keyData serialization.Data, reservationID int64) (*nearCacheRecord, error) {
+	key = rs.makeMapKey(key)
+	rs.recordsMu.Lock()
+	defer rs.recordsMu.Unlock()
+	rec, ok := rs.records[key]
+	if ok {
+		return rec, nil
+	}
+	rec, err := rs.createRecord(nil)
+	if err != nil {
+		return nil, err
+	}
+	rec.SetReservationID(reservationID)
+	rs.records[key] = rec
+	// initInvalidationMetaData(record, key, keyData);
+	return rec, nil
+}
+
+func (rs *nearCacheRecordStore) createRecord(value interface{}) (*nearCacheRecord, error) {
+	// assumes recordsMu was locked elsewhere
+	var err error
+	value, err = rs.valueConverter.ConvertValue(value)
+	if err != nil {
+		return nil, err
+	}
+	created := time.Now().UnixMilli()
+	expired := nearCacheRecordStoreTimeNotSet
+	if rs.timeToLiveMillis > 0 {
+		expired = created + rs.timeToLiveMillis
+	}
+	return newNearCacheRecord(value, created, expired), nil
+}
+
+func (rs *nearCacheRecordStore) updateRecordValue(rec *nearCacheRecord, value interface{}) error {
+	value, err := rs.valueConverter.ConvertValue(value)
+	if err != nil {
+		return err
+	}
+	rec.SetValue(value)
+	return nil
+}
+
+type nearCacheRecord struct {
+	CreationTime         int64
+	value                internal.AtomicValue
+	UUID                 types.UUID
+	PartitionID          int32
+	lastAccessTime       int64
+	ExpirationTime       int64
```

Comment:
These are int32 on Java client. I think int64 is more convenient for implementation but it would be great to make a note about it on TDD. Because I guess it is for memory optimization.

Reply:
int32 is the correct one. I think I've changed that in part 2, let me recheck...